src/components/ResourceList/ResourceList.tsx

```diff
@@ -80,86 +75,104 @@ export interface ResourceListProps {
   sortOptions?: SelectOption[];
   /** ReactNode to display instead of the sort control */
   alternateTool?: React.ReactNode;
+
   /** Callback when sort option is changed */
   onSortChange?(selected: string, id: string): void;
+
   /** Callback when selection is changed */
   onSelectionChange?(selectedItems: ResourceListSelectedItems): void;
-  /** Function to render each list item	 */
-  renderItem(item: any, id: string, index: number): React.ReactNode;
+  /** Function to render each list item   */
+  renderItem(item: ItemType, id: string, index: number): React.ReactNode;
   /** Function to customize the unique ID for each item */
-  idForItem?(item: any, index: number): string;
+  idForItem?(item: ItemType, index: number): string;
   /** Function to resolve an id from a item */
-  resolveItemId?(item: any): string;
+  resolveItemId?(item: ItemType): string;
 }
 
-type CombinedProps = ResourceListProps & WithAppProviderProps;
-
-class ResourceListInner extends React.Component<CombinedProps, State> {
-  static Item = ResourceItem;
+type ResourceListType = (<ItemType>(
+  value: ResourceListProps<ItemType>,
+) => ReactElement) & {
+  Item: typeof ResourceItem;
   // eslint-disable-next-line import/no-deprecated
-  static FilterControl = FilterControl;
+  FilterControl: typeof FilterControl;
+};
+
+export const ResourceList: ResourceListType = function ResourceList<ItemType>({
+  items,
+  filterControl,
+  resourceName: resourceNameProp,
+  promotedBulkActions,
+  bulkActions,
+  selectedItems = [],
+  selectable,
+  hasMoreItems,
+  loading,
+  showHeader = false,
+  totalItemsCount,
+  sortValue,
+  sortOptions,
+  alternateTool,
+  onSortChange,
+  onSelectionChange,
+  renderItem,
+  idForItem = (item: any, index: number) =>
+    Object.prototype.hasOwnProperty.call(item, 'id')
+      ? item.id
+      : index.toString(),
+  resolveItemId,
+}: ResourceListProps<ItemType>) {
+  const i18n = useI18n();
+  const [selectMode, setSelectMode] = useState(
+    Boolean(selectedItems && selectedItems.length > 0),
+  );
+  const [loadingPosition, setLoadingPositionState] = useState(0);
+  const [lastSelected, setLastSelected] = useState<number>();
+  const [smallScreen, setSmallScreen] = useState(isSmallScreen());
+
+  const [checkableButtons, setCheckableButtons] = useState<CheckableButtons>(
+    new Map(),
+  );
 
-  private defaultResourceName: {singular: string; plural: string};
-  private listRef: React.RefObject<HTMLUListElement> = React.createRef();
+  const defaultResourceName = useLazyRef(() => ({
```

Comment:
I am curious about the performance impacts here as well.

Reply:
As far as I know, it shouldn't have any performance impacts, our translations are very cheap. With that being said we also have no reason to translate the value multiple times. It's probably a case of why do it if you don't have too.

Reply:
That makes some sense to me. Are we happy with the LazyRef here?

Reply:
The impact is pretty light as Andrew said, but I'm trying to avoid unneeded function calls if possible

If we do the following:

```js
const defaultResourceName = useRef({
    singular: i18n.translate('Polaris.ResourceList.defaultItemSingular'),
    plural: i18n.translate('Polaris.ResourceList.defaultItemPlural'),
});
```

then that object is evaluated on every render of this component - we call ` i18n.translate` twice on every render of this component. By using `useLazyRef` then we compute this object on the first render of the component only. Thus avoiding these calls to ` i18n.translate` on any subsequent render.


 `const blah = useRef(someFunction())` then `someFunction` gets called every time this functional component is rendered.

Reply:
Cool, I am happy to leave the lazy ref there. Thanks for the detailed explanation!